[BUG] Integer Overflow occurs when searching a large-scale vector index with Memory Optimized Search

## Description
Creating a gh issue from forum question: https://forum.opensearch.org/t/integer-overflow-occurs-when-building-a-large-scale-vector-index/27818

## Issue Details

Versions (relevant - OpenSearch/Dashboard/Server OS/Browser):

OpenSearch 3.3.2

### Issue:

I built a vector index (faiss engine, HNSW + BQ, 1024 dimension per vector) with memory_optimized_search enabled (totally over 10 billion vectors) and the index setting is shown as follows:
```
"settings": {
   "index": {
       "knn": true,
       "knn.memory_optimized_search": true,
       "replication.type": "SEGMENT",
       "number_of_shards": 232,
        "number_of_replicas": 0
    }
}
```

After inserting all the data, I performed a force_merge to reduce segments to 1 per shard (approx. 43 million documents per shard). After that, when I searched the index with a query vector and an error “integer overflow” occurred

The full stack looks like:
```
Caused by: org.opensearch.core.common.io.stream.NotSerializableExceptionWrapper: arithmetic_exception: integer overflow
        at java.lang.Math.toIntExact(Math.java:1374) ~[?:?]
        at org.opensearch.knn.memoryoptsearch.faiss.MonotonicIntegerSequenceEncoder.encode(MonotonicIntegerSequenceEncoder.java:63) ~[?:?]
        at org.opensearch.knn.memoryoptsearch.faiss.FaissHNSW.load(FaissHNSW.java:77) ~[?:?]
        at org.opensearch.knn.memoryoptsearch.faiss.binary.FaissBinaryHnswIndex.doLoad(FaissBinaryHnswIndex.java:51) ~[?:?]
        at org.opensearch.knn.memoryoptsearch.faiss.FaissIndex.load(FaissIndex.java:55) ~[?:?]
        at org.opensearch.knn.memoryoptsearch.faiss.FaissIdMapIndex.doLoad(FaissIdMapIndex.java:58) ~[?:?]
        at org.opensearch.knn.memoryoptsearch.faiss.FaissIndex.load(FaissIndex.java:55) ~[?:?]
        at org.opensearch.knn.memoryoptsearch.faiss.FaissMemoryOptimizedSearcher.<init>(FaissMemoryOptimizedSearcher.java:49) ~[?:?]
        at org.opensearch.knn.memoryoptsearch.faiss.FaissMemoryOptimizedSearcherFactory.createVectorSearcher(FaissMemoryOptimizedSearcherFactory.java:39) ~[?:?]
        at org.opensearch.knn.index.codec.KNN990Codec.NativeEngines990KnnVectorsReader.lambda$getVectorSearcherSupplier$0(NativeEngines990KnnVectorsReader.java:372) ~[?:?]
        at org.opensearch.knn.index.codec.KNN990Codec.NativeEngines990KnnVectorsReader.loadMemoryOptimizedSearcherIfRequired(NativeEngines990KnnVectorsReader.java:325) ~[?:?]
        at org.opensearch.knn.index.codec.KNN990Codec.NativeEngines990KnnVectorsReader.trySearchWithMemoryOptimizedSearch(NativeEngines990KnnVectorsReader.java:252) ~[?:?]
        at org.opensearch.knn.index.codec.KNN990Codec.NativeEngines990KnnVectorsReader.search(NativeEngines990KnnVectorsReader.java:196) ~[?:?]
        at org.apache.lucene.codecs.perfield.PerFieldKnnVectorsFormat$FieldsReader.search(PerFieldKnnVectorsFormat.java:324) ~[lucene-core-10.3.1.jar:10.3.1 51190f35a16d2ce433139abfe0fd8365791b352a - 2025-10-02 09:50:16]
        at org.opensearch.knn.index.query.memoryoptsearch.MemoryOptimizedKNNWeight.queryIndex(MemoryOptimizedKNNWeight.java:202) ~[?:?]
        at org.opensearch.knn.index.query.memoryoptsearch.MemoryOptimizedKNNWeight.doANNSearch(MemoryOptimizedKNNWeight.java:100) ~[?:?]
        at org.opensearch.knn.index.query.KNNWeight.approximateSearch(KNNWeight.java:505) ~[?:?]
        at org.opensearch.knn.index.query.KNNWeight.searchLeaf(KNNWeight.java:336) ~[?:?]
        at org.opensearch.knn.index.query.nativelib.NativeEngineKnnVectorQuery.searchLeaf(NativeEngineKnnVectorQuery.java:438) ~[?:?]
        at org.opensearch.knn.index.query.nativelib.NativeEngineKnnVectorQuery.lambda$doSearch$0(NativeEngineKnnVectorQuery.java:272) ~[?:?]
```

The main issue is caused by MonotonicIntegerSequenceEncoder.encode function:
```
final long value = Math.toIntExact(input.readLong());
```

Is there any solution to this problem? It seems that it’s a bug when indexing large amounts of vectors with memory_optimized_search enabled.


After analyzing the code, I guess an individual FAISS HNSW segment cannot exceed `Integer.MAX_VALUE` vectors, otherwise, the error mentioned above will occur.

However, a segment only has 43 million vectors in my experiment. Based on the code, it should not produce this integer overflow error. Please correct me if I’m wrong.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -14,6 +14,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 * The KNN1030Codec does not properly support delegation for non-default codec(s). [#3093](https://github.com/opensearch-project/k-NN/pull/3093)
 * Fix score conversion logic for radial exact search [#3110](https://github.com/opensearch-project/k-NN/pull/3110)
 * Fix bugs in optimistic search for nested Cagra index [#3155](https://github.com/opensearch-project/k-NN/pull/3155)
+* Fix integer overflow for memory optimized search [#3130](https://github.com/opensearch-project/k-NN/pull/3130)
 
 ### Refactoring
 

--- a/src/main/java/org/opensearch/knn/memoryoptsearch/faiss/FaissHNSW.java
+++ b/src/main/java/org/opensearch/knn/memoryoptsearch/faiss/FaissHNSW.java
@@ -74,7 +74,7 @@ public class FaissHNSW {
 
         // Load `offsets` into memory.
         size = input.readLong();
-        offsetsReader = MonotonicIntegerSequenceEncoder.encode(Math.toIntExact(size), input);
+        offsetsReader = MonotonicIntegerSequenceEncoder.encode(Math.toIntExact(size), input, false);
         Objects.requireNonNull(offsetsReader);
 
         // Mark neighbor list section.

--- a/src/main/java/org/opensearch/knn/memoryoptsearch/faiss/MonotonicIntegerSequenceEncoder.java
+++ b/src/main/java/org/opensearch/knn/memoryoptsearch/faiss/MonotonicIntegerSequenceEncoder.java
@@ -33,6 +33,10 @@ public class MonotonicIntegerSequenceEncoder {
      * @throws IOException
      */
     public static DirectMonotonicReader encode(final int numElements, final IndexInput input) throws IOException {
+        return encode(numElements, input, true);
+    }
+
+    public static DirectMonotonicReader encode(final int numElements, final IndexInput input, final boolean isInteger) throws IOException {
         // Prepare a buffer for meta
         ByteBuffersDataOutput dataOutput = new ByteBuffersDataOutput();
         ByteBuffersIndexOutput dataIndexOutput = new ByteBuffersIndexOutput(
@@ -57,14 +61,12 @@ public class MonotonicIntegerSequenceEncoder {
             DIRECT_MONOTONIC_BLOCK_SHIFT
         );
 
-        // Encode integer sequence.
+        // Encode integer or long sequence.
         boolean isIdenticalMapping = true;
-        for (long i = 0; i < numElements; i++) {
-            final long value = Math.toIntExact(input.readLong());
-            if (value != i) {
-                isIdenticalMapping = false;
-            }
-            encoder.add(value);
+        if (isInteger) {
+            isIdenticalMapping = encodeInteger(encoder, numElements, input);
+        } else {
+            isIdenticalMapping = encodeLong(encoder, numElements, input);
         }
 
         encoder.finish();
@@ -91,5 +93,29 @@ public class MonotonicIntegerSequenceEncoder {
             DIRECT_MONOTONIC_BLOCK_SHIFT
         );
         return DirectMonotonicReader.getInstance(encodingMeta, dataInput);
+    }
+
+    private static boolean encodeInteger(DirectMonotonicWriter encoder, final int numElements, final IndexInput input) throws IOException {
+        boolean isIdenticalMapping = true;
+        for (long i = 0; i < numElements; i++) {
+            final long value = Math.toIntExact(input.readLong());
+            if (value != i) {
+                isIdenticalMapping = false;
+            }
+            encoder.add(value);
+        }
+        return isIdenticalMapping;
+    }
+
+    private static boolean encodeLong(DirectMonotonicWriter encoder, final int numElements, final IndexInput input) throws IOException {
+        boolean isIdenticalMapping = true;
+        for (long i = 0; i < numElements; i++) {
+            final long value = input.readLong();
+            if (value != i) {
+                isIdenticalMapping = false;
+            }
+            encoder.add(value);
+        }
+        return isIdenticalMapping;
     }
 }

--- a/src/test/java/org/opensearch/knn/memoryoptsearch/MonotonicIntegerSequenceEncoderTests.java
+++ b/src/test/java/org/opensearch/knn/memoryoptsearch/MonotonicIntegerSequenceEncoderTests.java
@@ -112,6 +112,24 @@ public class MonotonicIntegerSequenceEncoderTests extends KNNTestCase {
         }
     }
 
+    @SneakyThrows
+    public void testIncreasingSequenceWithLongEncoder() {
+        // Create a sequence
+        final int size = 100000;
+        final IndexInput input = createSequence((i) -> i / 5, size);
+
+        // Call encode
+        DirectMonotonicReader reader = MonotonicIntegerSequenceEncoder.encode(size, input, false);
+
+        // It should not be null
+        assertNotNull(reader);
+
+        // Validate values
+        for (int i = 0; i < size; ++i) {
+            assertEquals(i / 5, reader.get(i));
+        }
+    }
+
     private static IndexInput createSequence(Function<Long, Long> mapping, int length) throws IOException {
         ByteBuffersDataOutput dataOutput = new ByteBuffersDataOutput();
         ByteBuffersIndexOutput dataIndexOutput = new ByteBuffersIndexOutput(

